aten/src/ATen/core/boxing/impl/boxing.h

```diff
@@ -7,73 +7,252 @@
 template<class Result, class... Args>
-std::enable_if_t<supports_boxing<Result, Args...>::value && !std::is_same<void, Result>::value, Result>
-boxAndCallBoxedFunc(KernelFunction::InternalBoxedKernelFunction* boxed_kernel_func, OperatorKernel* functor, const OperatorHandle& opHandle, Args... args) {
-  // TODO Reuse stack vector instead of allocating?
-  torch::jit::Stack stack;
-  torch::jit::push(stack, std::forward<Args>(args)...);
+struct BoxedKernelWrapper<Result(Args...), std::enable_if_t<has_quantizer_arg<Args...>::value, void>> {
+  static Result call(KernelFunction::InternalBoxedKernelFunction*, OperatorKernel*, const OperatorHandle&, Args... args) {
+    TORCH_INTERNAL_ASSERT(false, "Unboxed call to a boxed kernel with unboxable parameter type at::Quantizer.");
+  }
+};
 
-  (*boxed_kernel_func)(functor, opHandle, &stack);
+// 2. Supported signatures, other than ref-passing.
+//
+template<class Result, class... Args>
+struct BoxedKernelWrapper<
+  Result(Args...),
+  std::enable_if_t<can_box_all<Args...>::value && can_unbox<Result>::value, void>
+> {
+  static Result call(
+    KernelFunction::InternalBoxedKernelFunction* boxed_kernel_func,
+    OperatorKernel* functor,
+    const OperatorHandle& opHandle,
+    Args... args
+  ) {
+    // TODO Reuse stack vector instead of allocating?
+    torch::jit::Stack stack;
+    torch::jit::push(stack, std::forward<Args>(args)...);
 
-  TORCH_INTERNAL_ASSERT(stack.size() == 1, "A boxed kernel should only push one return to the stack");
-  return std::move(stack[0]).to<Result>();
-}
+    (*boxed_kernel_func)(functor, opHandle, &stack);
+
+    return guts::if_constexpr<!std::is_same<void, Result>::value>(
+      [&] (auto delay_check) {
+        TORCH_INTERNAL_ASSERT(
+          stack.size() == 1,
+          "Boxed kernel was expected to push exactly one return value to the stack."
+        );
+        return delay_check(std::move(stack[0]).to<Result>());
+      },
+      [&] {
+        TORCH_INTERNAL_ASSERT(
+          stack.size() == 0,
+          "Boxed kernel for op with void return type pushed one or more return values to the stack."
+        );
+      }
+    );
+  }
+};
+
+// 3. signatures returning a single reference of the same type as
+// their initial argument.
+//
+// Note that the passed kernels are assumed to be for inplace/outplace ops,
+// and the generated BoxedKernelWrapper specializations will simply return
+// the initial argument.
+//
+template<class Result, class... OtherArgs>
+struct BoxedKernelWrapper<
+  Result(Result, OtherArgs...),
+  std::enable_if_t<
+    can_box_all<Result, OtherArgs...>::value && std::is_lvalue_reference<Result>::value,
+    void
+  >
+> {
+  static Result call(
+    KernelFunction::InternalBoxedKernelFunction* boxed_kernel_func,
+    OperatorKernel* functor,
+    const OperatorHandle& opHandle,
+    Result outArg,
+    OtherArgs... otherArgs
+  ) {
+    // TODO Reuse stack vector instead of allocating?
+    torch::jit::Stack stack;
```

Comment:
hm this has some non-trivial changes to the above version, so folding it in with `if_constexpr` might not make sense. If we can figure out how to do it, however, then you could get rid of all your explaining comments about which version does what, they're only there because SFINAE is generally hard to understand IIUC and an if/else might not need them